From 034196f6d61312d0cb218691d80c0e1ffadd4842 Mon Sep 17 00:00:00 2001
From: Lukas Bezdicka <lbezdick@redhat.com>
Date: Mon, 13 Jul 2015 17:39:39 +0200
Subject: [PATCH] Setup SELinux booleans if running in httpd

SELinux booleans httpd_use_openstack and httpd_can_network_connect_db
need to be turned on in order to have keystone working with httpd on
SELinux enabled systems.

Change-Id: I8722a4e619ac0d0a9e534e34721409d72e5c8323
---
 keystone/manifests/wsgi/apache.pp                  | 18 ++++++++++++++++++
 keystone/spec/classes/keystone_wsgi_apache_spec.rb | 13 ++++++++++++-
 2 files changed, 30 insertions(+), 1 deletion(-)

diff --git a/keystone/manifests/wsgi/apache.pp b/keystone/manifests/wsgi/apache.pp
index 080e28c..2e9bd1c 100644
--- a/keystone/manifests/wsgi/apache.pp
+++ b/keystone/manifests/wsgi/apache.pp
@@ -106,6 +106,9 @@
 #     (optional) Passes a string of custom configuration
 #     directives to be placed at the end of the vhost configuration.
 #     Defaults to undef.
+#   [*setup_selinux*]
+#     Setup SELinux ? (boolean)
+#     Optional. Defaults to true
 #
 # == Dependencies
 #
@@ -155,6 +158,7 @@ class keystone::wsgi::apache (
 
   $access_log_format     = false,
   $vhost_custom_fragment = undef,
+  $setup_selinux         = true,
 ) {
 
   include ::keystone::params
@@ -164,6 +168,20 @@ class keystone::wsgi::apache (
     include ::apache::mod::ssl
   }
 
+  if $setup_selinux and (str2bool($::selinux) == true) {
+    Selboolean <| |> ~> Service['httpd']
+    # we need httpd to connect to keystone ports
+    selboolean { 'httpd_use_openstack':
+      value      => 'on',
+      persistent => true,
+    }
+    # we need httpd to connect to databases
+    selboolean { 'httpd_can_network_connect_db':
+      value      => 'on',
+      persistent => true,
+    }
+  }
+
   Package['keystone'] -> Package['httpd']
   Package['keystone'] ~> Service['httpd']
   Keystone_config <| |> ~> Service['httpd']
diff --git a/keystone/spec/classes/keystone_wsgi_apache_spec.rb b/keystone/spec/classes/keystone_wsgi_apache_spec.rb
index bad9576..4e5263f 100644
--- a/keystone/spec/classes/keystone_wsgi_apache_spec.rb
+++ b/keystone/spec/classes/keystone_wsgi_apache_spec.rb
@@ -271,7 +271,8 @@ describe 'keystone::wsgi::apache' do
     let :facts do
       global_facts.merge({
         :osfamily               => 'RedHat',
-        :operatingsystemrelease => '6.0'
+        :operatingsystemrelease => '6.0',
+        :selinux                => 'true'
       })
     end
 
@@ -285,6 +286,16 @@ describe 'keystone::wsgi::apache' do
     end
 
     it_configures 'apache serving keystone with mod_wsgi'
+
+    it { is_expected.to contain_selboolean('httpd_use_openstack').with(
+      'value'      => 'on',
+      'persistent' => true,
+    )}
+
+    it { is_expected.to contain_selboolean('httpd_can_network_connect_db').with(
+      'value'      => 'on',
+      'persistent' => true,
+    )}
   end
 
   context 'on Debian platforms' do
